### DISCUSSION TITLE
ANSI colours with downlit

### DISCUSSION THREAD

**Discussion by mccarthy-m-g** [quarto-dev/quarto-cli#3717] [Q&A] [answered]
Does Quarto support ANSI colours in R output? This is handled by `downlit` in R packages like `pkgdown` and "just works". Quarto uses `downlit` for code linking when `code-link: true`, but this doesn't make colours work.

For example, in this reprex the output will just be plain text, but it should be coloured.

````
---
title: "Reprex"
format:
  html:
    code-link: true
---

```{r}
#| collapse = TRUE
cat(cli::col_red("This is red"), "\n")
cat(cli::col_blue("This is blue"), "\n")

message(cli::col_green("This is green"))

warning(cli::style_bold("This is bold"))
```
````

**Comment by cscheid** [comment]
I don't know the interaction between `knitr`, `downlit`, and `cli::col_*` well enough to say what's going wrong, but the markdown we're getting back from `knitr` doesn't have any color annotations in it (I got this by simply adding `keep-md: true` to your example):

````
---
title: "Reprex"
format:
  html:
    code-link: true
    keep-md: true
---


::: {.cell}

```{.r .cell-code}
cat(cli::col_red("This is red"), "\n")
## This is red
cat(cli::col_blue("This is blue"), "\n")
## This is blue

message(cli::col_green("This is green"))
## This is green

warning(cli::style_bold("This is bold"))
## Warning: This is bold
```
:::
````

**Reply by cscheid** [reply]
@cderv would you know how this all works by any chance?

**Comment by cderv** [comment, marked as answer]
> This is handled by downlit in R packages like pkgdown and "just works"

@mccarthy-m-g This is only working in **pkgdown** context right ? Or do you also get coloring working in R Markdown document ? 

**cli** does ANSI coloring, and not HTML coloring. So to make in work in a HTML rendered document, **downlit** is converting to HTML highlighting (with help of [`fansi`](https://github.com/brodieG/fansi) 📦. Moreover, I think by default **cli** function deactivate ANSI coloring in a knitting context. So it needs to be reactivated to be then handled by **downlit**. 

Also I believe **pkgdown** is doing a special post processing before **downlit** as ANSI escape character can be lost otherwise. 

All that above explain that this does not "just works" currently with Quarto. 

Can you confirm that it works for you only in **pkgdown** context ? From there, I'll know where to look to try improve this behavior. From my test, it does not work either in R Markdown, only in **pkgdown** special handling. 

By the **fansi** can be used directly to convert from ANSI with a knitr hooks.  See https://github.com/brodieG/fansi#rmarkdown
It applies with Quarto too. 

Example to illustrate how this works: 

````markdown
---
title: "Reprex"
format: html
---

```{r}
#| comment: ''
#| results: asis
#| echo: false
old_hook <- fansi::set_knit_hooks(knitr::knit_hooks, which = c("output", "warning", "error", "message"))
options(crayon.enabled = TRUE)
```

```{r}
cat(cli::col_red("This is red"), "\n")
cat(cli::col_blue("This is blue"), "\n")

message(cli::col_green("This is green"))

warning(cli::style_bold("This is bold"))
```
````

![image](https://user-images.githubusercontent.com/6791940/208689059-1ebb4f3c-87ef-4c13-80cf-c3c4e9fa66d2.png)


Hope it helps understand. But anyway, I'll think about how improving this with `code_link: true` as indeed **downlit** is supposed to handle the ANSI conversion to HTML and CSS.

**Reply by cscheid** [reply]
I knew you'd know everything about this inside and out 🙇‍♂️

**Reply by mccarthy-m-g** [reply]
This is excellent, thank you @cderv!

Just for reference in case anyone else is trying this, you can also set `options(cli.num_colors = 256)` if you need more than the 8 default colours.

@cscheid  Do you want me to file an issue re: supporting ANSI colours when `code_link: true`?

**Reply by cderv** [reply]
> Do you want me to file an issue re: supporting ANSI colours when code_link: true?

yes please do, this will be easier to follow. However, I feel this is something that should be dealt with in the R packages. 

> Can you confirm that it works for you only in pkgdown context ?

I am still interesting by this answer 😄

**Reply by mccarthy-m-g** [reply]
Done!

I tested with pkgdown, rmarkdown, and distill: Only pkgdown gives colour output natively.

**Reply by cderv** [reply]
Ok this confirm my understanding. We probably could try do something in **rmarkdown** natively when **downlit** is used.
